src/boot/efi/cpio.c

```diff
@@ -0,0 +1,465 @@
+        }
+
+        return pack_cpio_dir(path, dir_mode, inode_counter, cpio_buffer, cpio_buffer_size);
+}
+
+
+static EFI_STATUS pack_cpio_trailer(
+                VOID **cpio_buffer,
+                UINTN *cpio_buffer_size) {
+
+        static const char trailer[] =
+                "070701"
+                "00000000"
+                "00000000"
+                "00000000"
+                "00000000"
+                "00000001"
+                "00000000"
+                "00000000"
+                "00000000"
+                "00000000"
+                "00000000"
+                "00000000"
+                "0000000B"
+                "00000000"
+                "TRAILER!!!\0\0\0"; /* There's a fourth NUL byte appended here, because this is a string */
+
+        VOID *a;
+
+        /* Generates the cpio trailer record that indicates the end of our initrd cpio archive */
+
+        assert(cpio_buffer);
+        assert(cpio_buffer_size);
+        assert_cc(sizeof(trailer) % 4 == 0);
+
+        a = ReallocatePool(*cpio_buffer, *cpio_buffer_size, *cpio_buffer_size + sizeof(trailer));
+        if (!a)
+                return EFI_OUT_OF_RESOURCES;
+
+        *cpio_buffer = a;
+        CopyMem((UINT8*) *cpio_buffer + *cpio_buffer_size, trailer, sizeof(trailer));
+        *cpio_buffer_size += sizeof(trailer);
+
+        return EFI_SUCCESS;
+}
+
+EFI_STATUS pack_cpio(
+                EFI_LOADED_IMAGE *loaded_image,
+                const CHAR16 *match_suffix,
+                const CHAR8 *target_dir_prefix,
+                UINT32 dir_mode,
+                UINT32 access_mode,
+                UINTN tpm_pcr,
+                const CHAR16 *tpm_description,
+                VOID **ret_buffer,
+                UINTN *ret_buffer_size) {
+
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE root = NULL, extra_dir = NULL;
+        UINTN dirent_size = 0, buffer_size = 0, l, a,  n_items = 0, n_allocated = 0;
+        _cleanup_freepool_ EFI_FILE_INFO *file_info = NULL, *dirent = NULL;
+        _cleanup_freepool_ CHAR16 *loaded_image_path = NULL, *j = NULL;
+        _cleanup_(strv_freep) CHAR16 **items = NULL;
+        _cleanup_freepool_ VOID *buffer = NULL;
+        UINT32 inode = 1; /* inode counter, so that each item gets a new inode */
+        EFI_STATUS err;
+
+        assert(loaded_image);
+        assert(ret_buffer);
+        assert(ret_buffer_size);
+
+        root = LibOpenRoot(loaded_image->DeviceHandle);
+        if (!root)
+                return log_error_status_stall(EFI_LOAD_ERROR, L"Unable to open root directory.");
+
+        loaded_image_path = DevicePathToStr(loaded_image->FilePath);
+        if (!loaded_image_path)
+                return log_oom();
+
+        l = StrLen(loaded_image_path);
+        a = StrLen(EXTRA_DIR_SUFFIX);
+        if (l >= UINTN_MAX - a)
+                return log_error_status_stall(EFI_OUT_OF_RESOURCES, L"Path length overflow.");
+
+        j = AllocatePool(sizeof(CHAR16) * (l + a + 1));
```

Comment:
Technically, this could still overflow since it does not use sizeof(CHAR16) above, no? You could just use StrSize() for these (I think they're only used here).